build: Use nextflow gradle plugin

edmundmiller opened the pull request:
This pull request refactors the build and plugin management system for the project, simplifying the `Makefile`, removing the root-level `plugins/build.gradle`, and updating Gradle settings for plugin management. The changes streamline how plugins are built, tested, installed, and released, and shift plugin configuration to a more modern Gradle setup.

Follows https://www.nextflow.io/docs/latest/guides/migrate-plugin.html

edmundmiller committed:
build: Use nextflow gradle plugin

Signed-off-by: Edmund Miller <edmund.miller@seqera.io>

bentsherman commented on build.gradle:
```diff
@@ -14,19 +14,34 @@
 
-idea {
-    module.inheritOutputDirs = true
+sourceSets {
+    main {
+        groovy {
+            srcDirs = ['src/main/groovy']
+        }
+        resources {
+            srcDirs = ['src/main/resources']
+        }
+    }
+    test {
+        groovy {
+            srcDirs = ['src/test/groovy']
+        }
+    }
 }
```
I think this can be removed

bentsherman commented on build.gradle:
```diff
@@ -36,75 +51,86 @@ repositories {
-    // JDBC driver setup for AWS Athena - the 3rd party JAR are being downloaded and setup as gradle tasks below.
-    // Reference https://docs.aws.amazon.com/athena/latest/ug/connect-with-jdbc.html
+    
+    // Replace commons-logging with jcl-over-slf4j
+    implementation 'org.slf4j:jcl-over-slf4j:2.0.7'
```
I think this can be removed

bentsherman commented on build.gradle:
```diff
@@ -36,75 +51,86 @@ repositories {
-    testImplementation "io.nextflow:nextflow:$nextflowVersion"
     testImplementation ("org.apache.groovy:groovy-test:4.0.23") { exclude group: 'org.apache.groovy' }
     testImplementation ("cglib:cglib-nodep:3.3.0")
     testImplementation ("org.objenesis:objenesis:3.1")
     testImplementation ("org.spockframework:spock-core:2.3-groovy-4.0") { exclude group: 'org.apache.groovy'; exclude group: 'net.bytebuddy' }
     testImplementation ('org.spockframework:spock-junit4:2.3-groovy-4.0') { exclude group: 'org.apache.groovy'; exclude group: 'net.bytebuddy' }
     testImplementation ('com.google.jimfs:jimfs:1.1')
 
-    testImplementation(testFixtures("io.nextflow:nextflow:$nextflowVersion"))
-    testImplementation(testFixtures("io.nextflow:nf-commons:$nextflowVersion"))
+    // Add back Nextflow test dependencies
+    testImplementation "io.nextflow:nextflow:${nextflowPlugin.nextflowVersion}"
+    testImplementation(testFixtures("io.nextflow:nextflow:${nextflowPlugin.nextflowVersion}"))
+    testImplementation(testFixtures("io.nextflow:nf-commons:${nextflowPlugin.nextflowVersion}"))
```
I think all of this can be removed

bentsherman commented on build.gradle:
```diff
@@ -36,75 +51,86 @@ repositories {
+    testImplementation(testFixtures("io.nextflow:nf-commons:${nextflowPlugin.nextflowVersion}"))
+}
+
+configurations {
+    // see https://docs.gradle.org/4.1/userguide/dependency_management.html#sub:exclude_transitive_dependencies
+    runtimeClasspath.exclude group: 'org.slf4j', module: 'slf4j-api'
+}
```
I think this can be removed

bentsherman commented on build.gradle:
```diff
@@ -36,75 +51,86 @@ repositories {
- * for Simba Athena JDBC driver and extract its contents to the build directory
- * Reference: https://docs.aws.amazon.com/athena/latest/ug/connect-with-jdbc.html
- */
+// AWS Athena dependencies setup - migrated from the original build.gradle
+apply plugin: 'de.undercouch.download'
```
This can be moved to the `plugins` block at the top

bentsherman commented on settings.gradle:
```diff
@@ -1,3 +1,11 @@
+// For the Nextflow plugin Gradle plugin
+pluginManagement {
+    repositories {
+        mavenLocal()
+        gradlePluginPortal()
+    }
+}
```
This should not be needed anymore since the nextflow gradle plugin was published to Maven Central

pditommaso reviewed (commented):
Please keep this on hold, until we don't have a pre-production plugin and registry

pditommaso commented:
Worth to resume this now that the registry is avail

edmundmiller committed:
Address reviewer feedback and simplify build configuration

- Remove unnecessary sourceSets configuration per @bentsherman feedback
- Remove explicit test dependencies (handled by plugin)
- Remove SLF4J and runtime classpath exclusions
- Simplify settings.gradle (remove pluginManagement block)
- Keep only essential database dependencies and Athena tasks
- Update to latest Nextflow Gradle plugin v1.0.0-beta.6
- Maintain full functionality while reducing complexity

All tests pass and plugin builds/installs successfully.

Co-authored-by: Ben Sherman <bentsherman@users.noreply.github.com>
Signed-off-by: Edmund Miller <edmund.miller@seqera.io>
